Objects/rangeobject.c

```diff
@@ -527,35 +527,101 @@ range_count(rangeobject *r, PyObject *ob)
 }
 
 static PyObject *
-range_index(rangeobject *r, PyObject *ob)
+range_index(rangeobject *r, PyObject *args, PyObject *kw)
 {
     int contains;
+    static char *keywords[] = {"value", "start", "stop", NULL};
+    PyObject *ob;               /* ob is the key */
+    PyObject *start = NULL;
+    PyObject *stop = NULL;
+    if(!PyArg_ParseTupleAndKeywords(args, kw, "O|OO:range_index",
+                                    keywords, &ob, &start, &stop)){
+      return NULL;
+    }
 
+    PyObject *return_value;
+    if(start || stop){
+      /* This means we have to use only a portion of the range. We
+         create a relavant slice of the range and do the indexing on
+         that slice */
+        if(start){
+            if(!PyLong_CheckExact(start)){
+                PyErr_Format(PyExc_TypeError,
+                             "start value must be integer, not %.200s",
+                             start->ob_type->tp_name);
+                return NULL;
+            }
+        }
+        if(stop){
+            if(!PyLong_CheckExact(stop)){
+                PyErr_Format(PyExc_TypeError,
+                             "stop value must be integer, not %.200s",
+                             stop->ob_type->tp_name);
+                return NULL;
+            }
+        }
+        if(!start){
+            start = _PyLong_Zero;
+        }
+        if(!stop){
+            stop = r->length;
+        }
+        PyObject *slice = PySlice_New(start, stop, _PyLong_One);
+        r = (rangeobject*) compute_slice(r, slice);
+        Py_DECREF(slice);
```

Comment:
Slice object creation maybe unsuccessful in which case the return value maybe NULL. Therefore use Py_XDECREF. 

```suggestion
        Py_XDECREF(slice);
```